File: transaction_parser/parser_benchmark/report/transaction_parser_benchmark_log_detail/transaction_parser_benchmark_log_detail.py
# Copyright (c) 2026, Resilient Tech and contributors
# For license information, please see license.txt

from collections import defaultdict

import frappe
from frappe import _
from frappe.query_builder.functions import Coalesce

PARTY_TYPE_MAP = {
    "Sales Order": "Customer",
    "Expense": "Supplier",
}


def execute(filters=None):
    filters = frappe._dict(filters or {})
    report = BenchmarkLogDetailReport(filters)
    return report.run()


class BenchmarkLogDetailReport:
    def __init__(self, filters: frappe._dict):
        self.filters = filters
        self._set_party_type()

    def run(self):
        logs = self._fetch_logs()
        if not logs:
            return self._get_columns(), []

        log_names = [r.log_name for r in logs]
        score_details_map = self._fetch_score_details(log_names)

        data = [self._build_row(r, score_details_map) for r in logs]

        # discover all unique key names for dynamic columns
        all_keys = dict.fromkeys(
            k for row in data for k in (row.pop("_key_accuracies", None) or {})
        )

        return self._get_columns(list(all_keys)), data

    # ── Columns ──────────────────────────────────────────────────────

    def _get_columns(self, key_names=None):
        columns = [
            {
                "fieldname": "log_name",
                "label": _("Log"),
                "fieldtype": "Link",
                "options": "Parser Benchmark Log",
                "width": 160,
            },
            {
                "fieldname": "dataset",
                "label": _("Dataset"),
                "fieldtype": "Link",
                "options": "Parser Benchmark Dataset",
                "width": 140,
            },
            {
                "fieldname": "party",
                "label": _("Party"),
                "fieldtype": "Data",
                "width": 160,
            },
            {
                "fieldname": "ai_model",
                "label": _("AI Model"),
                "fieldtype": "Data",
                "width": 170,
            },
            {
                "fieldname": "pdf_processor",
                "label": _("Processor"),
                "fieldtype": "Data",
                "width": 100,
            },
            {
                "fieldname": "accuracy_score",
                "label": _("Accuracy (%)"),
                "fieldtype": "Percent",
                "width": 115,
            },
            {
                "fieldname": "status",
                "label": _("Status"),
                "fieldtype": "Data",
                "width": 90,
            },
            {
                "fieldname": "commit_hash",
                "label": _("Commit"),
                "fieldtype": "Data",
                "width": 90,
            },
            {
                "fieldname": "commit_message",
                "label": _("Commit Message"),
                "fieldtype": "Data",
                "width": 200,
            },
            {
                "fieldname": "creation",
                "label": _("Date"),
                "fieldtype": "Datetime",
                "width": 160,
            },
            {
                "fieldname": "total_time",
                "label": _("Total (s)"),
                "fieldtype": "Float",
                "width": 90,
                "precision": 2,
            },
            {
                "fieldname": "total_cost",
                "label": _("Cost"),
                "fieldtype": "Currency",
                "width": 90,
                "options": "currency",
            },
            {
                "fieldname": "total_tokens",
                "label": _("Tokens"),
                "fieldtype": "Int",
                "width": 85,
            },
        ]

        # dynamic per-key accuracy columns
        for key in key_names or []:
            columns.append(
                {
                    "fieldname": f"key_{key}",
                    "label": _(key.replace("_", " ").title() + " (%)"),
                    "fieldtype": "Percent",
                    "width": 120,
                }
            )

        return columns

    # ── Query ────────────────────────────────────────────────────────

    def _fetch_logs(self):
        log = frappe.qb.DocType("Parser Benchmark Log")
        ds = frappe.qb.DocType("Parser Benchmark Dataset")
        cust = frappe.qb.DocType("Customer")
        supp = frappe.qb.DocType("Supplier")

        statuses = ["Completed"]
        if self.filters.get("include_failed"):
            statuses.append("Failed")

        query = (
            frappe.qb.from_(log)
            .join(ds)
            .on(log.dataset == ds.name)
            .left_join(cust)
            .on((ds.party_type == "Customer") & (ds.party == cust.name))
            .left_join(supp)
            .on((ds.party_type == "Supplier") & (ds.party == supp.name))
            .select(
                log.name.as_("log_name"),
                log.creation,
                log.status,
                log.ai_model,
                log.pdf_processor,
                log.accuracy_score,
                log.total_time,
                log.total_cost,
                log.total_tokens,
                log.currency,
                log.dataset,
                log.commit_hash,
                log.commit_message,
                ds.party,
                Coalesce(cust.customer_name, supp.supplier_name, ds.party).as_(
                    "party_name"
                ),
            )
            .where(log.status.isin(statuses))
            .where(ds.docstatus == 1)
            .orderby(log.creation, order=frappe.qb.desc)
        )

        if not self.filters.get("include_disabled_datasets"):
            query = query.where(ds.enabled == 1)

        if self.filters.get("is_multiple_files"):
            query = query.where(ds.is_multiple_files == 1)

        for column, key in (
            (ds.company, "company"),
            (ds.transaction_type, "transaction_type"),
            (ds.party_type, "party_type"),
            (ds.party, "party"),
            (log.dataset, "dataset"),
        ):
            if self.filters.get(key):
                query = query.where(column == self.filters[key])

        for column, key in (
            (log.ai_model, "ai_model"),
            (log.pdf_processor, "pdf_processor"),
        ):
            values = self.filters.get(key)
            if values:
                items = values if isinstance(values, list) else [values]
                query = query.where(column.isin(items))

        return query.run(as_dict=True)

    def _fetch_score_details(self, log_names: list[str]) -> dict[str, list[dict]]:
        if not log_names:
            return {}

        sd = frappe.qb.DocType("Parser Benchmark Score Detail")
        rows = (
            frappe.qb.from_(sd)
            .select(sd.parent, sd.key, sd.accuracy)
            .where(sd.parent.isin(log_names))
            .orderby(sd.idx)
            .run(as_dict=True)
        )

        details_map: dict[str, list[dict]] = defaultdict(list)
        for row in rows:
            details_map[row.parent].append(row)

        return details_map

    # ── Helpers ──────────────────────────────────────────────────────

    def _set_party_type(self):
        transaction_type = self.filters.get("transaction_type")
        if transaction_type and not self.filters.get("party_type"):
            self.filters["party_type"] = PARTY_TYPE_MAP.get(transaction_type)

    def _build_row(self, r, score_details_map):
        details = score_details_map.get(r.log_name, [])
        key_accuracies = {d["key"]: d["accuracy"] for d in details}

        commit_msg = (r.commit_message or "").split("\n")[0][:80]

        row = {
            "log_name": r.log_name,
            "dataset": r.dataset,
            "party": r.party_name or r.party or _("No Party"),
            "ai_model": r.ai_model,
            "pdf_processor": r.pdf_processor,
            "accuracy_score": r.accuracy_score if details else "",
            "status": r.status,
            "commit_hash": (r.commit_hash or "")[:7],
            "commit_message": commit_msg,
            "creation": r.creation,
            "total_time": r.total_time,
            "total_cost": r.total_cost,
            "total_tokens": r.total_tokens,
            "currency": r.currency,
            "_key_accuracies": key_accuracies,
        }

        for k, v in key_accuracies.items():
            row[f"key_{k}"] = v

        return row

File: transaction_parser/parser_benchmark/report/transaction_parser_benchmark_log_detail/transaction_parser_benchmark_log_detail.js
// Copyright (c) 2026, Resilient Tech and contributors
// For license information, please see license.txt

const AI_MODELS = [
	"DeepSeek Chat",
	"DeepSeek Reasoner",
	"OpenAI gpt-4o",
	"OpenAI gpt-4o-mini",
	"OpenAI gpt-5",
	"OpenAI gpt-5-mini",
	"Google Gemini Pro-2.5",
	"Google Gemini Flash-2.5",
	"Claude Haiku-4.5",
];

const PDF_PROCESSORS = ["OCRMyPDF", "Docling"];

const PARTY_TYPE_MAP = {
	"Sales Order": "Customer",
	Expense: "Supplier",
};

function make_options(items, txt) {
	return items
		.filter((v) => !txt || v.toLowerCase().includes(txt.toLowerCase()))
		.map((v) => ({ value: v, description: "" }));
}

frappe.query_reports["Transaction Parser Benchmark Log Detail"] = {
	onload(report) {
		set_party_type(report);
	},

	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "transaction_type",
			label: __("Transaction Type"),
			fieldtype: "Select",
			options: "\nSales Order\nExpense",
			reqd: 1,
			default: "Sales Order",
			on_change() {
				set_party_type(frappe.query_report);
			},
		},
		{
			fieldname: "party_type",
			label: __("Party Type"),
			fieldtype: "Link",
			options: "DocType",
			hidden: 1,
		},
		{
			fieldname: "party",
			label: __("Party"),
			fieldtype: "Dynamic Link",
			options: "party_type",
		},
		{
			fieldname: "dataset",
			label: __("Dataset"),
			fieldtype: "Link",
			options: "Parser Benchmark Dataset",
			get_query() {
				const filters = {};
				const party_type = frappe.query_report.get_filter_value("party_type");
				const party = frappe.query_report.get_filter_value("party");

				if (party_type) filters.party_type = party_type;
				if (party) filters.party = party;

				return { filters };
			},
		},
		{
			fieldname: "ai_model",
			label: __("AI Model"),
			fieldtype: "MultiSelectList",
			get_data: (txt) => make_options(AI_MODELS, txt),
		},
		{
			fieldname: "pdf_processor",
			label: __("PDF Processor"),
			fieldtype: "MultiSelectList",
			get_data: (txt) => make_options(PDF_PROCESSORS, txt),
		},
		{
			fieldname: "include_disabled_datasets",
			label: __("Include Disabled Datasets"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "is_multiple_files",
			label: __("Multiple Files Only"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "include_failed",
			label: __("Include Failed"),
			fieldtype: "Check",
			default: 0,
		},
	],
};

function set_party_type(report) {
	const transaction_type = report.get_filter_value("transaction_type");
	const party_type = PARTY_TYPE_MAP[transaction_type] || "";
	report.set_filter_value("party_type", party_type);
}
